Sofa/Component/Engine/Select/src/sofa/component/engine/select/BaseROI.h

```diff
@@ -0,0 +1,177 @@
+* along with this program. If not, see <http://www.gnu.org/licenses/>.        *
+*******************************************************************************
+* Authors: The SOFA Team and external contributors (see Authors.txt)          *
+*                                                                             *
+* Contact information: contact@sofa-framework.org                             *
+******************************************************************************/
+#pragma once
+#include <sofa/component/engine/select/config.h>
+
+#include <sofa/type/Vec.h>
+#include <sofa/core/DataEngine.h>
+#include <sofa/core/objectmodel/BaseObject.h>
+#include <sofa/core/behavior/MechanicalState.h>
+#include <sofa/core/topology/BaseMeshTopology.h>
+#include <sofa/defaulttype/RigidTypes.h>
+
+namespace sofa::component::engine::select
+{
+
+/**
+ * This interface defines the features for Regions Of Interest (ROI)
+ */
+template <class DataTypes>
+class BaseROI : public core::DataEngine
+{
+public:
+    using Inherit = core::DataEngine;
+    SOFA_CLASS(SOFA_TEMPLATE(BaseROI,DataTypes), Inherit);
+
+    using VecCoord = VecCoord_t<DataTypes>;
+    using Coord = Coord_t<DataTypes>;
+    using Real = Real_t<DataTypes>;
+    using CPos = typename DataTypes::CPos;
+
+    using PointID = core::topology::BaseMeshTopology::PointID;
+    using SetIndex = core::topology::BaseMeshTopology::SetIndex;
+    using Edge = core::topology::BaseMeshTopology::Edge;
+    using VecEdge = type::vector<Edge>;
+    using Triangle = core::topology::BaseMeshTopology::Triangle;
+    using VecTriangle = type::vector<Triangle>;
+    using Quad = core::topology::BaseMeshTopology::Quad;
+    using VecQuad = type::vector<Quad>;
+    using Tetra = core::topology::BaseMeshTopology::Tetra;
+    using VecTetra = type::vector<Tetra>;
+    using Hexa = core::topology::BaseMeshTopology::Hexa;
+    using VecHexa = type::vector<Hexa>;
+
+public:
+    void init() final;
+    void doUpdate() override; // Cannot set final because of the tests (e.g. DataEngine_test)
+    void draw(const core::visual::VisualParams* vparams) final; 
+    void computeBBox(const core::ExecParams* params, bool onlyVisible) final;
+
+    virtual void roiInit() {};
+    virtual bool roiDoUpdate() { return true; };
+    virtual void roiDraw(const core::visual::VisualParams*) {};
+    virtual void roiComputeBBox(const core::ExecParams*, type::BoundingBox&) {};
+
+    /// Pre-construction check method called by ObjectFactory.
+    /// Check that DataTypes matches the MechanicalState.
+    template<class T>
+    static bool canCreate(T*& obj, core::objectmodel::BaseContext* context, core::objectmodel::BaseObjectDescription* arg)
+    {
+        if (!arg->getAttribute("template"))
+        {
+            // only check if this template is correct if no template was given
+            if (context->getMechanicalState() && dynamic_cast<core::behavior::MechanicalState<DataTypes>*>(context->getMechanicalState()) == nullptr)
```

Comment:
I find it strange to rely on a MechanicalState whereas it is possible to use this component without it. I mean, the position vector can be set by the user, or by a MeshLoader. There is even not a permanent Link to any MechanicalState.

Reply:
True that.
I guess it predates the code in the init(), data linking, dataengine, etc
I will remove it